File: src/routes/score-profile/score-profile.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  Query,
  HttpCode,
  HttpStatus,
  ParseIntPipe,
  ValidationPipe,
} from '@nestjs/common'
import { ApiTags, ApiOperation, ApiResponse, ApiParam, ApiQuery } from '@nestjs/swagger'
import { ScoreProfileService } from './score-profile.service'
import {
  CreateScoreProfileDto,
  UpdateScoreProfileDto,
  ScoreProfileQueryDto,
  ScoreProfileResponseDto,
  ScoreProfileListResponseDto,
} from './score-profile.dto'

@ApiTags('score-profiles')
@Controller('score-profiles')
export class ScoreProfileController {
  constructor(private readonly scoreProfileService: ScoreProfileService) {}

  @Post()
  async create(@Body(ValidationPipe) createDto: CreateScoreProfileDto): Promise<ScoreProfileResponseDto> {
    return this.scoreProfileService.create(createDto)
  }

  @Get()
  @ApiOperation({ summary: 'Get all score profiles with pagination' })
  @ApiQuery({ name: 'page', required: false, example: 1 })
  @ApiQuery({ name: 'limit', required: false, example: 20 })
  @ApiQuery({ name: 'level', required: false, enum: ['N5', 'N4', 'N3', 'N2', 'N1'] })
  @ApiQuery({ name: 'name', required: false })
  @ApiQuery({ name: 'sortBy', required: false, enum: ['id', 'name', 'level', 'createdAt', 'updatedAt'] })
  @ApiQuery({ name: 'sortOrder', required: false, enum: ['asc', 'desc'] })
  async findMany(@Query(ValidationPipe) query: ScoreProfileQueryDto): Promise<ScoreProfileListResponseDto> {
    return this.scoreProfileService.findMany(query)
  }

  @Get(':id')
  async findById(@Param('id', ParseIntPipe) id: number): Promise<ScoreProfileResponseDto> {
    return this.scoreProfileService.findById(id)
  }

  @Put(':id')
  async update(
    @Param('id', ParseIntPipe) id: number,
    @Body(ValidationPipe) updateDto: UpdateScoreProfileDto,
  ): Promise<ScoreProfileResponseDto> {
    return this.scoreProfileService.update(id, updateDto)
  }

  @Delete(':id')
  @HttpCode(HttpStatus.OK)
  async delete(@Param('id', ParseIntPipe) id: number) {
    return this.scoreProfileService.delete(id)
  }
}

File: assessment-service/src/main/java/com/torii/assessment/controller/ScoreProfileController.java
package com.torii.assessment.controller;

import com.torii.assessment.dto.scoreprofile.*;
import com.torii.assessment.service.ScoreProfileService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/score-profiles")
@RequiredArgsConstructor
public class ScoreProfileController {
    
    private final ScoreProfileService scoreProfileService;
    
    @PostMapping
    public ResponseEntity<ScoreProfileResponseDTO> create(@Valid @RequestBody CreateScoreProfileDTO dto) {
        ScoreProfileResponseDTO response = scoreProfileService.create(dto);
        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }
    
    @GetMapping
    public ResponseEntity<ScoreProfileListResponseDTO> findMany(
            @RequestParam(required = false, defaultValue = "1") Integer page,
            @RequestParam(required = false, defaultValue = "20") Integer limit,
            @RequestParam(required = false) String level,
            @RequestParam(required = false) String name,
            @RequestParam(required = false, defaultValue = "createdAt") String sortBy,
            @RequestParam(required = false, defaultValue = "desc") String sortOrder) {
        
        ScoreProfileQueryDTO query = new ScoreProfileQueryDTO();
        query.setPage(page);
        query.setLimit(limit);
        query.setLevel(level);
        query.setName(name);
        query.setSortBy(sortBy);
        query.setSortOrder(sortOrder);
        
        ScoreProfileListResponseDTO response = scoreProfileService.findMany(query);
        return ResponseEntity.ok(response);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<ScoreProfileResponseDTO> findById(@PathVariable Long id) {
        ScoreProfileResponseDTO response = scoreProfileService.findById(id);
        return ResponseEntity.ok(response);
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<ScoreProfileResponseDTO> update(
            @PathVariable Long id,
            @Valid @RequestBody UpdateScoreProfileDTO dto) {
        ScoreProfileResponseDTO response = scoreProfileService.update(id, dto);
        return ResponseEntity.ok(response);
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<Map<String, String>> delete(@PathVariable Long id) {
        scoreProfileService.delete(id);
        return ResponseEntity.ok(Map.of("message", "Score profile deleted successfully"));
    }
}
